File: backend/app/api/v1/reports.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.models.session_notes import SessionNotes
from app.models.session import Session as SessionModel
from app.models.feedback import DailyFeedback
from app.crud import child as crud_child
from app.api.deps import get_current_user

router = APIRouter()


@router.get("/progress/{child_id}")
async def get_progress_report(
    child_id: str,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    child = crud_child.get_child_by_id(db, child_id)
    if not child:
        return {"error": "Child not found"}

    sessions = (
        db.query(SessionModel)
        .filter(SessionModel.child_id == child_id)
        .order_by(SessionModel.scheduled_at.desc())
        .limit(10)
        .all()
    )

    session_ids = [s.id for s in sessions]
    notes = db.query(SessionNotes).filter(SessionNotes.session_id.in_(session_ids)).all()

    feedbacks = (
        db.query(DailyFeedback)
        .filter(DailyFeedback.child_id == child_id)
        .order_by(DailyFeedback.feedback_date.desc())
        .limit(30)
        .all()
    )

    avg_scores = {}
    if notes:
        score_fields = ["attention_score", "participation_score", "mood_score", "progress_score", "behavior_score"]
        for field in score_fields:
            values = [getattr(n, field) for n in notes if getattr(n, field) is not None]
            avg_scores[field] = round(sum(values) / len(values), 2) if values else None

    sessions_completed = sum(1 for s in sessions if s.status.value == "completed")
    sessions_scheduled = len(sessions)
    attendance_percent = round(sessions_completed / sessions_scheduled * 100) if sessions_scheduled > 0 else 0

    all_score_values = [v for v in avg_scores.values() if v is not None]
    average_progress = round(sum(all_score_values) / len(all_score_values), 2) if all_score_values else 0.0

    progress_trend = [
        {"label": f"W{i + 1}", "score": float(n.progress_score or 0)}
        for i, n in enumerate(reversed(notes[:8]))
    ]

    therapist_note = notes[0].next_session_plan if notes and notes[0].next_session_plan else None

    return {
        "child_id": str(child_id),
        "child_name": child.name,
        "sessions_completed": sessions_completed,
        "sessions_scheduled": sessions_scheduled,
        "attendance_percent": attendance_percent,
        "average_progress": average_progress,
        "average_therapy_scores": avg_scores,
        "total_feedback_entries": len(feedbacks),
        "progress_trend": progress_trend,
        "therapist_note": therapist_note,
    }

File: backend-node/src/routes/reports.js
const express = require('express');
const db = require('../db');
const { authenticate } = require('../middleware/auth');

const router = express.Router();

router.get('/progress/:childId', authenticate, async (req, res) => {
  const { childId } = req.params;

  const childRes = await db.query('SELECT * FROM children WHERE id = $1', [childId]);
  if (!childRes.rows.length) return res.json({ error: 'Child not found' });
  const child = childRes.rows[0];

  const sessionsRes = await db.query(
    `SELECT * FROM sessions WHERE child_id = $1 ORDER BY scheduled_at DESC LIMIT 10`,
    [childId]
  );
  const sessions = sessionsRes.rows;
  const sessionIds = sessions.map(s => s.id);

  let notes = [];
  if (sessionIds.length) {
    const notesRes = await db.query(
      `SELECT * FROM session_notes WHERE session_id = ANY($1)`,
      [sessionIds]
    );
    notes = notesRes.rows;
  }

  const feedbackRes = await db.query(
    `SELECT * FROM daily_feedback WHERE child_id = $1
     ORDER BY feedback_date DESC LIMIT 30`,
    [childId]
  );
  const feedbacks = feedbackRes.rows;

  const scoreFields = ['attention_score', 'participation_score', 'mood_score',
                       'progress_score', 'behavior_score'];
  const avgScores = {};
  for (const field of scoreFields) {
    const vals = notes.map(n => n[field]).filter(v => v != null);
    avgScores[field] = vals.length
      ? Math.round(vals.reduce((a, b) => a + b, 0) / vals.length * 100) / 100
      : null;
  }

  const sessionsCompleted = sessions.filter(s => s.status === 'completed').length;
  const sessionsScheduled = sessions.length;
  const attendancePercent = sessionsScheduled > 0
    ? Math.round(sessionsCompleted / sessionsScheduled * 100) : 0;

  const allVals = Object.values(avgScores).filter(v => v != null);
  const averageProgress = allVals.length
    ? Math.round(allVals.reduce((a, b) => a + b, 0) / allVals.length * 100) / 100 : 0;

  const progressTrend = [...notes].reverse().slice(0, 8).map((n, i) => ({
    label: `W${i + 1}`,
    score: parseFloat(n.progress_score || 0),
  }));

  const therapistNote = notes.length && notes[0].next_session_plan
    ? notes[0].next_session_plan : null;

  res.json({
    child_id: childId,
    child_name: child.name,
    sessions_completed: sessionsCompleted,
    sessions_scheduled: sessionsScheduled,
    attendance_percent: attendancePercent,
    average_progress: averageProgress,
    average_therapy_scores: avgScores,
    total_feedback_entries: feedbacks.length,
    progress_trend: progressTrend,
    therapist_note: therapistNote,
  });
});

module.exports = router;